### PR TITLE
Bug 1946929: Fix status reporting for default node placement

### PR DESCRIPTION
**Pull request by Miciah**
#### `getOperatorState`: Delete parameter and simplify

* `pkg/operator/controller/status/controller.go` (`Reconcile`): Delete argument to `getOperatorState`.
(`getOperatorState`): Delete namespace name parameter.  Use `DefaultOperandNamespace` for the namespace name.
* `test/e2e/operator_test.go` (`TestClusterOperatorStatusRelatedObjects`): Add a comment.  Log the observed `status.relatedObjects` value when it doesn't match the expected value.  Use `%q` to log the clusteroperator name and `%w` where possible to report error values.

#### `operatorState`: Make fields private

* `pkg/operator/controller/status/controller.go` (`Reconcile`, `operatorState`, `getOperatorState`): Make `operatorState` fields private.

#### status: Don't assume dns exists

When updating the clusteroperator's status, check if a dns exists at all, and properly handle the case where it does not.

* `pkg/operator/controller/status/controller.go` (`Reconcile`): Use the new `haveNamespace` field from `operatorState` to determine if the operand namespace exists.  Pass the new `haveDNS` field as well as the `dns` field from `operatorState` to `computeOperatorStatusVersions`, `computeOperatorAvailableCondition`, `computeOperatorProgressingCondition`, and `computeOperatorDegradedCondition`.
(`operatorState`): Add `haveNamespace` and `haveDNS` fields.  Change the type of the `namespace` field from a pointer type to a non-pointer type.
(`getOperatorState`): Delete now-unnecessary temporary variable.  Check if at least 1 dns exists.  Set the `haveNamespace` and `haveDNS` fields of `operatorState`.
(`computeOperatorStatusVersions`): Replace the `allDNSesAvailable` parameter with `haveDNS` and `dns` parameters.  Return the old versions if there is no dns or the dns reports `Available=False`.
(`computeOperatorDegradedCondition`): Replace the `allDNSesAvailable` parameter with `haveDNS` and `dns` parameters.  Report `Degraded=True` if there is no dns or the dns reports `Degraded=True`.
(`computeOperatorDegradedCondition`): Replace the `dnsAvailable` parameter with `haveDNS` and `dns` parameters.  Report `Progressing=True` if there is no dns or the dns reports `Available=False`.
(`computeOperatorAvailableCondition`): Replace the `dnsAvailable` parameter with `haveDNS` and `dns` parameters.  Report `Available=False` if there is no dns or the dns reports `Available=False`.
* `pkg/operator/controller/status/controller_test.go` (`TestComputeOperatorProgressingCondition`): Add "dns does not exist" test case.  Update call to `computeOperatorProgressingCondition` to pass in a dns when appropriate.
(`TestComputeOperatorStatusVersions`): Add "initialize versions, dns does not exist" test case.  Update call to `computeOperatorStatusVersions` to pass in a dns when appropriate.

#### status: Compute operator's Progressing from dns's

Change the computation of the clusteroperator "Progressing" status condition to reflect the dns's "Progressing" status condition instead of its "Available" status condition.

* `pkg/operator/controller/status/controller.go` (`computeOperatorProgressingCondition`): Check the dns's `Progressing` status condition, report `Progressing=True` if the dns reports `Progressing=True`, and report `Progressing=Unknown` if the dns reports `Progressing=Unknown` or doesn't report a `Progressing` status condition.
* `pkg/operator/controller/status/controller_test.go` (`TestComputeOperatorProgressingCondition`): Add test cases to verify that `computeOperatorProgressingCondition` accurately reflects the dns's Progressing status condition.

#### status: Don't update versions if still progressing

Change the computation of the clusteroperator versions status to continue reporting the previously reported versions as long as the dns reports "Progressing=True".  The versions should only be updated once the dns is done progressing.

* `pkg/operator/controller/status/controller.go` (`computeOperatorStatusVersions`): Check the dns's `Progressing` status condition instead of its `Available` status condition.
* `pkg/operator/controller/status/controller_test.go` (`TestComputeOperatorStatusVersions`): Fix test cases to use the `Progressing` status condition.

#### status/controller_test: Change `Fatalf` to `Errorf`

Use `Errorf` when iterating over test cases so that all the test cases are run (and errors printed for all failing test cases) instead of stopping on the first test case to fail.

* `pkg/operator/controller/status/controller_test.go` (`TestComputeOperatorProgressingCondition`, `TestOperatorStatusesEqual`, `TestComputeOperatorStatusVersions`): Replace `Fatalf` with `Errorf`.

#### status: Combine status condition merging

* `pkg/operator/controller/status/controller.go` (`Reconcile`): Replace 3 separate invocations of `mergeConditions` with a single invocation.

#### Fix status reporting for default node placement

Fix the computation of the dns's "Progressing" status condition to report correct status when the default node selector or tolerations are used. Before this change, the dns reported `Progressing=True` when no node selector or no tolerations was specified.

Improve the readability of the status condition message when reporting that the daemonset does not have the expected node selector or tolerations.

This change fixes [bug 1946929](https://bugzilla.redhat.com/show_bug.cgi?id=1946929).

* `assets/dns/daemonset.yaml`: Delete tolerations and add a comment that tolerations is set at run-time.
* `pkg/manifests/bindata.go`: Regenerate.
* `pkg/operator/controller/controller_dns_daemonset.go` (`desiredDNSDaemonSet`): Factor out the defaulting logic for the node selector and tolerations from here...
(`nodeSelectorForDNS`, `tolerationsForDNS`): ...to here.  New functions for computing the node selector and tolerations for DNS pods.
* `pkg/operator/controller/dns_status.go` (`computeDNSProgressingCondition`): Use the new `nodeSelectorForDNS` and `tolerationsForDNS` functions to determine what the desired node selector and tolerations are for comparison with the actual node selector and tolerations.  Replace `%v` with `%+v` when printing selectors and tolerations to improve readability.
* `pkg/operator/controller/dns_status_test.go` (`TestDNSStatusConditions`): Use the new `nodeSelectorForDNS` and `tolerationsForDNS` functions.
(`TestComputeDNSProgressingCondition`): Update test cases to use the default node selector and tolerations, and add explicit test cases where a custom node selector or tolerations is specified.

#### status: Group utilclock import properly

* `pkg/operator/controller/status/controller.go`: Group the `k8s.io/apimachinery/pkg/util/clock` import with the other `k8s.io/apimachinery` imports.

#### status: Delete superfluous print statements

* `pkg/operator/controller/status/controller.go` (`Reconcile`, `getOperatorState`): Delete print statements for error values that are returned to callers.  If logging were needed, logrus should be used.  However, logging and then returning an error value is redundant.

#### status: Clean up logging and initialization

* `pkg/operator/controller/status/controller.go` (`Reconcile`): Use `%q` for the clusteroperator name and `%w` for wrapped error values when returning errors. Move initialization of the clusteroperator name from here...
(`initializeClusterOperator`): ...to here.

#### `daemonset.yaml`: Replace `nodeSelector` with comment

Since the dns daemonset's node selector is always set at runtime now, delete `spec.template.spec.nodeSelector` in the dns daemonset yaml definition, and add a comment that it is set at runtime.

Follow-up to https://github.com/openshift/cluster-dns-operator/pull/209/commits/0a6056a5fea0b174e4ccf240cb549b0465b37c3b.

* `assets/dns/daemonset.yaml`: Replace `nodeSelector` stanza with a comment that the `nodeSelector` is set at runtime.
* `pkg/manifests/bindata.go`: Regenerate.

### CODE REVIEW

File: pkg/operator/controller/status/controller.go [hunk 1]
```diff
@@ -282,7 +295,10 @@ func computeOperatorProgressingCondition(dnsAvailable bool, oldVersions, curVers
 	progressing := false
 
 	var messages []string
-	if !dnsAvailable {
+	if !haveDNS {
+		messages = append(messages, `DNS "default" does not exist`)
+		progressing = true
+	} else if !checkDNSAvailable(dns) {
 		messages = append(messages, "DNS default unavailable")
```
Comment (sgreene570) [inline pkg/operator/controller/status/controller.go]: Should this message be `'DNS "default" unavailable` for consistency?

File: test/e2e/operator_test.go [hunk 2]
```diff
@@ -96,6 +96,8 @@ func TestOperatorAvailable(t *testing.T) {
 	}
 }
 
+// TestClusterOperatorStatusRelatedObjects verifies that the dns clusteroperator
```
Comment (sgreene570) [inline test/e2e/operator_test.go:99]: Thanks for taking the time to fixup this test!

File: pkg/operator/controller/status/controller.go [hunk 3]
```diff
@@ -270,21 +298,42 @@ func computeOperatorDegradedCondition(dns *operatorv1.DNS) configv1.ClusterOpera
 }
 
 // computeOperatorProgressingCondition computes the operator's current Progressing status state.
-func computeOperatorProgressingCondition(dnsAvailable bool, oldVersions, curVersions []configv1.OperandVersion,
-	operatorReleaseVersion, coreDNSImage, openshiftCLIImage, kubeRBACProxyImage string) configv1.ClusterOperatorStatusCondition {
+func computeOperatorProgressingCondition(haveDNS bool, dns *operatorv1.DNS, oldVersions, curVersions []configv1.OperandVersion, operatorReleaseVersion, coreDNSImage, openshiftCLIImage, kubeRBACProxyImage string) configv1.ClusterOperatorStatusCondition {
 	// TODO: Update progressingCondition when an ingresscontroller
```
Comment (sgreene570) [inline pkg/operator/controller/status/controller.go:302]: While we are at it, should we rewrite this comment? Or can it be just be removed? It doesn't make sense to me.

File: pkg/operator/controller/status/controller.go [hunk 4]
```diff
@@ -299,55 +348,63 @@ func computeOperatorProgressingCondition(dnsAvailable bool, oldVersions, curVers
...
-		availableCondition.Status = configv1.ConditionTrue
-		availableCondition.Reason = "AsExpected"
-		availableCondition.Message = "DNS default is available"
-	} else {
+
+	switch {
+	case !haveDNS:
+		availableCondition.Status = configv1.ConditionFalse
+		availableCondition.Reason = "DNSDoesNotExist"
+		availableCondition.Message = `DNS "default" does not exist.`
```
Comment (sgreene570) [inline pkg/operator/controller/status/controller.go:399]: Should we consider using `dns.Name` in place of `"default"`?

File: pkg/operator/controller/status/controller.go [hunk 5]
```diff
@@ -176,44 +179,60 @@ func initializeClusterOperator(co *configv1.ClusterOperator) {
...
+	}
+	if err := r.client.Get(context.TODO(), name, ns); err != nil {
 		if !errors.IsNotFound(err) {
-			fmt.Printf("failed to get ns %s: %v\n", nsName, err)
-			return state, fmt.Errorf("failed to get namespace %q: %v", nsName, err)
+			return state, fmt.Errorf("failed to get namespace %q: %w", name, err)
 		}
 	} else {
-		state.Namespace = ns
+		state.haveNamespace = true
```
Comment (sgreene570) [inline pkg/operator/controller/status/controller.go:202]: I might be missing the point of this, but would it make sense to set `state.haveNamspace` to false after line 200?

Comment (Miciah) [reply pkg/operator/controller/status/controller.go:203]: The zero value of `state.haveNamespace` is false.  I can add an assignment to reduce the risk of confusion.

File: pkg/operator/controller/status/controller.go (hunk 4 above)
Comment (Miciah) [reply pkg/operator/controller/status/controller.go:397]: In this case (`case !haveDNS`), `dns` is nil.  I could use `dns.Name` in the other cases, but the operator doesn't reconcile any dns other than "default" anyway, so it seemed just as well to use a string literal throughout.

File: pkg/operator/controller/status/controller.go (hunk 3 above)
Comment (Miciah) [reply pkg/operator/controller/status/controller.go]: Sorry, yeah, I keep meaning to remove this comment.  It appears to have been copied and pasted from the ingress operator.  I'll remove it.

**Comment by Miciah** [comment]
failed to acquire lease: status 503 Service Unavailable, status code 503

/test e2e-aws-operator

File: pkg/operator/controller/status/controller.go (hunk 4 above)
Comment (sgreene570) [reply pkg/operator/controller/status/controller.go:397]: Fair enough.

File: pkg/operator/controller/status/controller.go (hunk 5 above)
Comment (sgreene570) [reply pkg/operator/controller/status/controller.go:203]: > The zero value of `state.haveNamespace` is false. I can add an assignment to reduce the risk of confusion.

Right. I was more so concerned about the scenario where the operand namespace is somehow deleted after `state.haveNamespace` is set to true (although ideally that would never happen, I suppose).

Comment (Miciah) [reply pkg/operator/controller/status/controller.go:203]: `getOperatorState` declares a new `var state operatorState` every invocation, so it doesn't matter if a previous `getOperatorState` invocation returned `state` with `state.haveNamespace` set to true.

File: pkg/operator/controller/status/controller.go [hunk 6]
```diff
@@ -270,21 +298,39 @@ func computeOperatorDegradedCondition(dns *operatorv1.DNS) configv1.ClusterOpera
...
+				progressingReasons = append(progressingReasons, "DNSReportsProgressingIsTrue")
+				messages = append(messages, fmt.Sprintf("DNS %q reports Progressing=True: %q", dns.Name, cond.Message))
+			case operatorv1.ConditionUnknown:
+				progressingReasons = append(progressingReasons, "DNSReportsProgressingIsUnknown")
+				messages = append(messages, fmt.Sprintf("DNS %q reports Progressing=Unknown: %q", dns.Name, cond.Message))
+			}
+			break
+		}
+		if !foundProgressingCondition {
+			progressingReasons = append(progressingReasons, "DNSDoesNotReportsProgressingStatus")
```
Comment (sgreene570) [inline pkg/operator/controller/status/controller.go:331]:
```suggestion
			progressingReasons = append(progressingReasons, "DNSDoesNotReportsProgressingStatus")
```

Is `reports` a typo?

File: assets/dns/daemonset.yaml [hunk 7]
```diff
@@ -82,9 +82,7 @@ spec:
             path: Corefile
       - name: metrics-tls
         # secretName is set at runtime
-      tolerations:
-      - key: "node-role.kubernetes.io/master"
-        operator: Exists
+      # tolerations is set at runtime.
```
Comment (sgreene570) [inline assets/dns/daemonset.yaml:85]: Would you mind removing the hard-coded `nodeSelector` in `daemonset.yaml` and adding a comment similar to this one in it's place?

Comment (Miciah) [reply assets/dns/daemonset.yaml:84]: Done.

File: pkg/operator/controller/status/controller.go (hunk 6 above)
Comment (Miciah) [reply pkg/operator/controller/status/controller.go]: Fixed.  Thanks!

**Comment by Miciah** [comment]
Rebased.

**Comment by Miciah** [comment]
For the record, CI was all green before the last rebase.

**Comment by Miciah** [comment]
Looks like issues with the API:

    [sig-network] pods should successfully create sandboxes by other
    4 failures to create the sandbox
    ns/openshift-dns pod/dns-default-5mdgr node/ip-10-0-155-142.ec2.internal - never deleted - reason/FailedCreatePodSandBox Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_dns-default-5mdgr_openshift-dns_01654220-663a-4e91-bfeb-68e8d6d40e29_0(39c6ed594e9f9cb9747eb1cc7a96f8303cf0e20cec9456bd85f86c3655b0309d): Multus: [openshift-dns/dns-default-5mdgr]: error getting pod: Get "https://[api-int.ci-op-fnqwclrb-230ac.origin-ci-int-aws.dev.rhcloud.com]:6443/api/v1/namespaces/openshift-dns/pods/dns-default-5mdgr?timeout=1m0s": dial tcp 10.0.226.138:6443: connect: connection refused
    ns/e2e-k8s-service-lb-available-5133 pod/service-test-2qtjb node/ip-10-0-155-142.ec2.internal - never deleted - reason/FailedCreatePodSandBox Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_service-test-2qtjb_e2e-k8s-service-lb-available-5133_b6b67e65-9033-47e0-b486-2659e62446d9_0(faeb23d0eb30691ab0f4147996c25aa8380714c6901185bbb8be70500ebc7874): Multus: [e2e-k8s-service-lb-available-5133/service-test-2qtjb]: error getting pod: Get "https://[api-int.ci-op-fnqwclrb-230ac.origin-ci-int-aws.dev.rhcloud.com]:6443/api/v1/namespaces/e2e-k8s-service-lb-available-5133/pods/service-test-2qtjb?timeout=1m0s": dial tcp 10.0.226.138:6443: connect: connection refused
    ns/e2e-k8s-service-lb-available-5133 pod/service-test-2qtjb node/ip-10-0-155-142.ec2.internal - never deleted - reason/FailedCreatePodSandBox Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_service-test-2qtjb_e2e-k8s-service-lb-available-5133_b6b67e65-9033-47e0-b486-2659e62446d9_0(81cadee8707d1502ae6b97eb7fd98f91f2cedb7463e47a7c71db6816f4153081): Multus: [e2e-k8s-service-lb-available-5133/service-test-2qtjb]: error getting pod: Get "https://[api-int.ci-op-fnqwclrb-230ac.origin-ci-int-aws.dev.rhcloud.com]:6443/api/v1/namespaces/e2e-k8s-service-lb-available-5133/pods/service-test-2qtjb?timeout=1m0s": dial tcp 10.0.226.138:6443: connect: connection refused
    ns/loki pod/loki-promtail-58mk6 node/ip-10-0-244-75.ec2.internal - never deleted - reason/FailedCreatePodSandBox Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_loki-promtail-58mk6_loki_44433214-32fd-4ea5-be1c-9642dd80f9be_0(d1ccf08a1a11b22ae676f459615f6453876d74aa5c0ea7f2acaf38be72343c66): Multus: [loki/loki-promtail-58mk6]: error getting pod: Get "https://[api-int.ci-op-fnqwclrb-230ac.origin-ci-int-aws.dev.rhcloud.com]:6443/api/v1/namespaces/loki/pods/loki-promtail-58mk6?timeout=1m0s": dial tcp 10.0.226.138:6443: connect: connection refused

/test e2e-upgrade

**Comment by Miciah** [comment]
/hold cancel
The PR has been open for review for a few days; let's go ahead and get it merged to unblock further PRs.

**Pull request merged**